packages/@aws-cdk/core/lib/private/metadata-resource.ts

```diff
@@ -76,17 +47,90 @@ function makeCdkMetadataAvailableCondition() {
+ * and then all of the construct-versions are grouped with common prefixes together, grouping common parts in '{}' and separating items with ','.
+ *
+ * Example:
+ * [1.90.0!aws-cdk-lib.Stack, 1.90.0!aws-cdk-lib.Construct, 1.90.0!aws-cdk-lib.service.Resource, 0.42.1!aws-cdk-lib-experiments.NewStuff]
+ * Becomes:
+ * 1.90.0!aws-cdk-lib.{Stack,Construct,service.Resource},0.42.1!aws-cdk-lib-experiments.NewStuff
+ *
+ * The whole thing is then either included directly as plaintext as:
+ * v2:plaintext:{prefixEncodedList}
+ * Or is compressed and base64-encoded, and then formatted as:
+ * v2:deflate64:{prefixEncodedListCompressedAndEncoded}
+ *
+ * Exported/visible (and `forcePlaintext` parameter) for ease of testing.
+ */
+export function formatAnalytics(infos: ConstructInfo[], forcePlaintext: boolean = false) {
+  const fqnsByVersion = infos.reduce(function (grouped, info) {
+    (grouped[info.version] = grouped[info.version] ?? new Set()).add(info.fqn);
+    return grouped;
+  }, {} as Record<string, Set<string>>);
 
-  for (const key of Object.keys(runtime.libraries).sort()) {
-    modules.push(`${key}=${runtime.libraries[key]}`);
-  }
-  return modules.join(',');
-}
\ No newline at end of file
+  const plaintextEncodedConstructs = Object.entries(fqnsByVersion).map(([version, fqns]) => {
+    const versionTrie = new Trie();
+    [...fqns].forEach(fqn => insertFqnInTrie(fqn, versionTrie));
+    return `${version}!${prefixEncodeTrie(versionTrie)}`;
+  }).join(',');
+
+  const compressedConstructs = zlib.gzipSync(Buffer.from(plaintextEncodedConstructs)).toString('base64');
+
+  return (plaintextEncodedConstructs.length < compressedConstructs.length || forcePlaintext)
+    ? `v2:plaintext:${plaintextEncodedConstructs}`
+    : `v2:deflate64:${compressedConstructs}`;
+}
+
+/**
+ * Splits after non-alphanumeric characters (e.g., '.', '/') in the FQN
+ * and insert each piece of the FQN in nested map (i.e., simple trie).
+ */
+function insertFqnInTrie(fqn: string, treeRef: Trie) {
+  fqn.replace(/[^a-z0-9]/gi, '$& ').split(' ').forEach(fqnPart => {
+    const nextLevelTreeRef = treeRef.get(fqnPart) ?? new Trie();
+    treeRef.set(fqnPart, nextLevelTreeRef);
+    treeRef = nextLevelTreeRef;
+  });
+  return treeRef;
+}
+
+/**
+ * Prefix-encodes a "trie-ish" structure, using '{}' to group and ',' to separate siblings.
+ *
+ * Example input:
+ * ABC,ABD,AEF
+ *
+ * Example trie:
+ * A --> B --> C
+ *  |     \--> D
+ *  \--> E --> F
+ *
+ * Becomes:
+ * A{B{C,D},EF}
+ */
+function prefixEncodeTrie(trie: Trie) {
+  let prefixEncoded = '';
+  let isFirstEntryAtLevel = true;
+  [...trie.entries()].forEach(([key, value]) => {
+    if (!isFirstEntryAtLevel) {
+      prefixEncoded += ',';
+    }
+    isFirstEntryAtLevel = false;
+    prefixEncoded += key;
+    if (value.size > 1) {
+      prefixEncoded += '{';
+      prefixEncoded += prefixEncodeTrie(value);
+      prefixEncoded += '}';
+    } else if (value.size == 1) {
```

Comment:
If `value.size == 0` you probably shouldn't have inserted the `,`  either, right?

So either this `if` is unnecessary (because you have an implicit precondition on maps not being empty), or you're missing an `if` at the top of the loop body.

Reply:
This `if` is really just a minor optimization (skipping an unnecessary recursive call). It can be removed without changing funcitonal behavior. If `value.size === 0` then the call is made, `trie.entries` is empty, and the function does nothing.

This doesn't have a direct impact on whether or not to insert the `,`; different logical levels of the trie. The `key` is going to be inserted after the `,`; the size of the values/map just determines if underneath/past the key there are other deeper entries to insert.

I can remove the redundant `if` if you think it reads clearer without.